File: integration/bindings/client-js/runtime-bridge.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { AsyncLocalStorage } from "node:async_hooks";
import { createRequire } from "node:module";
import { execFile } from "node:child_process";
import { mkdtemp, readFile, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { Bridge } from "../../../packages/client-js/bridge.mts";

// The SDK Bridge over the Rust-owned client runtime (#134): request routing,
// callback transactions, wake/drain dispatch and lifecycle, on the real
// Node carrier.
const native = createRequire(import.meta.url)(
  "../../../bindings/node/axton-node.node",
);
const schemaUrl = new URL(
  "../../../fixtures/schemas/entry.json",
  import.meta.url,
);
const schema = JSON.parse(await readFile(schemaUrl, "utf8"));
const envelopes = JSON.parse(
  await readFile(
    new URL("../../../fixtures/bridge/envelopes.json", import.meta.url),
    "utf8",
  ),
);
const key = (id) => ({ model: "Entry", identity: { id } });
const create = (id, text) => ({
  kind: "direct",
  operation: {
    model: "Entry",
    op: "create",
    identity: { id },
    values: { text },
  },
});
const update = (id, text) => ({
  kind: "direct",
  operation: {
    model: "Entry",
    op: "update",
    identity: { id },
    values: { text },
  },
});
const read = (id) => ({ kind: "read", key: key(id) });
const deferred = () => {
  let resolve;
  const promise = new Promise((r) => (resolve = r));
  return { promise, resolve };
};
/** Whether `promise` is still pending after the actor had time to run it. */
async function pending(promise, ms = 50) {
  let timer;
  const marker = Symbol();
  try {
    return (
      (await Promise.race([
        promise.then(
          () => undefined,
          () => undefined,
        ),
        new Promise(
          (resolve) => (timer = setTimeout(() => resolve(marker), ms)),
        ),
      ])) === marker
    );
  } finally {
    clearTimeout(timer);
  }
}
async function withBridge(body, carrier = native) {
  const directory = await mkdtemp(join(tmpdir(), "axton-bridge-"));
  const { bridge, opened } = await Bridge.open(carrier, {
    path: join(directory, "client.sqlite"),
    schema,
  });
  try {
    await body(bridge, opened);
  } finally {
    await bridge.close();
    await rm(directory, { recursive: true, force: true });
  }
}

test("open answers the client id and schema state", async () => {
  await withBridge(async (bridge, opened) => {
    assert.equal(typeof opened.clientId, "string");
    assert.deepEqual(opened.schema, {
      rebuilt: false,
      pending: null,
      lastRebuild: null,
    });
    assert.equal(bridge.closed, false);
  });
});

test("overlapping tasks return to the correct waiter while a callback owns the transaction", async () => {
  await withBridge(async (bridge) => {
    await bridge.task(create("e", "A"));
    const order = [];
    const entered = deferred();
    const inside = deferred();
    const gate = deferred();
    const transaction = bridge.transaction(async (transactionId) => {
      entered.resolve(transactionId);
      await bridge.transactionCommand(
        transactionId,
        undefined,
        update("e", "B"),
      );
      order.push("tx write");
      const row = await bridge.transactionCommand(
        transactionId,
        undefined,
        read("e"),
      );
      order.push(`tx read ${row.text}`);
      inside.resolve();
      await gate.promise;
    });
    void transaction.then(() => order.push("transaction"));
    const transactionId = await entered.promise;
    assert.equal(typeof transactionId, "string");
    const outside = bridge.task(read("e")).then((row) => {
      order.push(`outside read ${row.text}`);
      return row;
    });
    const status = bridge.task({ kind: "status" }).then((state) => {
      order.push("status");
      return state;
    });
    await inside.promise;
    assert.equal(await pending(outside), true, "ordinary read waits");
    assert.equal(await pending(status), true, "status waits");
    assert.deepEqual(order, ["tx write", "tx read B"]);
    gate.resolve();
    await transaction;
    assert.equal((await outside).text, "B");
    assert.equal(typeof (await status).clientId, "string");
    assert.deepEqual(order, [
      "tx write",
      "tx read B",
      "transaction",
      "outside read B",
      "status",
    ]);
  });
});

test("a command in the wrong scope fails without joining and rolls the unit back", async () => {
  await withBridge(async (bridge) => {
    let refused;
    await assert.rejects(
      bridge.transaction(async (transactionId) => {
        await bridge.transactionCommand(
          transactionId,
          undefined,
          create("w", "A"),
        );
        refused = await bridge
          .transactionCommand(transactionId, "sp999", read("w"))
          .catch((error) => error);
      }),
      /invalid transaction scope/,
    );
    assert.match(refused.message, /invalid transaction scope/);
    assert.equal(await bridge.task(read("w")), null);
  });
});

test("savepoint scopes are issued by the runtime and carried by nested commands", async () => {
  await withBridge(async (bridge) => {
    await bridge.transaction(async (transactionId) => {
      const { scope } = await bridge.transactionCommand(
        transactionId,
        undefined,
        {
          kind: "savepoint",
        },
      );
      assert.equal(typeof scope, "string");
      await bridge.transactionCommand(transactionId, scope, create("s", "A"));
      await bridge.transactionCommand(transactionId, scope, {
        kind: "rollbackSavepoint",
        scope,
      });
      await bridge.transactionCommand(
        transactionId,
        undefined,
        create("t", "B"),
      );
    });
    assert.equal(await bridge.task(read("s")), null);
    assert.equal((await bridge.task(read("t"))).text, "B");
  });
});

test("admission failure rejects its waiter and leaves the bridge usable; after close tasks reject client_closed", async () => {
  let refuse = true;
  const carrier = {
    runtimeOpen: (request, wake) => native.runtimeOpen(request, wake),
    runtimeSubmit(runtimeId, message) {
      if (refuse && JSON.parse(message).command?.kind === "status") {
        refuse = false;
        throw Error("client_closed");
      }
      native.runtimeSubmit(runtimeId, message);
    },
    runtimeDrain: (runtimeId) => native.runtimeDrain(runtimeId),
    runtimeDetach: (runtimeId) => native.runtimeDetach(runtimeId),
  };
  let closedBridge;
  await withBridge(async (bridge) => {
    closedBridge = bridge;
    await assert.rejects(bridge.task({ kind: "status" }), /client_closed/);
    assert.equal(
      typeof (await bridge.task({ kind: "status" })).clientId,
      "string",
    );
  }, carrier);
  assert.equal(closedBridge.closed, true);
  await assert.rejects(closedBridge.task({ kind: "status" }), /client_closed/);
  await closedBridge.close();
});

test("a throwing callback rejects with the same value and rolls back", async () => {
  await withBridge(async (bridge) => {
    const thrown = Error("abort");
    await assert.rejects(
      bridge.transaction(async (transactionId) => {
        await bridge.transactionCommand(
          transactionId,
          undefined,
          create("rolled", "A"),
        );
        throw thrown;
      }),
      (error) => error === thrown,
    );
    assert.equal(await bridge.task(read("rolled")), null);
    const odd = { reason: "not an Error" };
    await assert.rejects(
      bridge.transaction(async () => {
        throw odd;
      }),
      (error) => error === odd,
    );
  });
});

test("the callback runs in the async context of the transaction's caller", async () => {
  await withBridge(async (bridge) => {
    const storage = new AsyncLocalStorage();
    let seen;
    await storage.run("caller", () =>
      bridge.transaction(async () => {
        seen = storage.getStore();
      }),
    );
    assert.equal(seen, "caller");
  });
});

test("a thrown value without a usable message still rolls back and rejects with it", async () => {
  await withBridge(async (bridge) => {
    const opaque = Object.create(null);
    await assert.rejects(
      bridge.transaction(async (transactionId) => {
        await bridge.transactionCommand(
          transactionId,
          undefined,
          create("o", "A"),
        );
        throw opaque;
      }),
      (error) => error === opaque,
    );
    assert.equal(await bridge.task(read("o")), null);
  });
});

test("a caught command failure rejects at commit with the engine message", async () => {
  await withBridge(async (bridge) => {
    let caught;
    await assert.rejects(
      bridge.transaction(async (transactionId) => {
        await bridge.transactionCommand(
          transactionId,
          undefined,
          create("kept", "A"),
        );
        await bridge
          .transactionCommand(transactionId, undefined, {
            kind: "direct",
            operation: {
              model: "Missing",
              op: "create",
              identity: { id: "m" },
              values: {},
            },
          })
          .catch((error) => {
            caught = error;
          });
      }),
      (error) =>
        error instanceof Error &&
        error !== caught &&
        error.message === caught.message,
    );
    assert.ok(caught instanceof Error);
    assert.equal(await bridge.task(read("kept")), null);
  });
});

test("a failed open rejects with the engine message and detaches its runtime", async () => {
  const directory = await mkdtemp(join(tmpdir(), "axton-bridge-open-"));
  const opened = [];
  const detached = [];
  const carrier = {
    runtimeOpen(request, wake) {
      const id = native.runtimeOpen(request, wake);
      opened.push(id);
      return id;
    },
    runtimeSubmit: (runtimeId, message) =>
      native.runtimeSubmit(runtimeId, message),
    runtimeDrain: (runtimeId) => native.runtimeDrain(runtimeId),
    runtimeDetach(runtimeId) {
      detached.push(runtimeId);
      native.runtimeDetach(runtimeId);
    },
  };
  try {
    const missing = join(directory, "missing", "deeper", "client.sqlite");
    await assert.rejects(
      Bridge.open(carrier, { path: missing, schema }),
      (error) =>
        error instanceof Error &&
        error.message.length > 0 &&
        error.message !== "client_closed",
    );
    assert.equal(opened.length, 1);
    assert.deepEqual(detached, opened);
    assert.throws(
      () => native.runtimeSubmit(opened[0], JSON.stringify({ type: "close" })),
      /client_closed/,
    );
    const { bridge } = await Bridge.open(carrier, {
      path: join(directory, "client.sqlite"),
      schema,
    });
    assert.equal(await bridge.task(read("none")), null);
    await bridge.close();
    assert.deepEqual(detached, opened);
  } finally {
    await rm(directory, { recursive: true, force: true });
  }
});

test("close during an open callback rolls back and settles every waiter", async () => {
  await withBridge(async (bridge) => {
    const entered = deferred();
    const gate = deferred();
    let late;
    const transaction = bridge.transaction(async (transactionId) => {
      await bridge.transactionCommand(
        transactionId,
        undefined,
        create("c", "A"),
      );
      entered.resolve(transactionId);
      await gate.promise;
      late = await bridge
        .transactionCommand(transactionId, undefined, read("c"))
        .catch((error) => error);
    });
    await entered.promise;
    const queued = bridge.task(read("c"));
    const closing = bridge.close();
    assert.strictEqual(bridge.close(), closing);
    await assert.rejects(transaction, /client_closed/);
    await assert.rejects(queued, /client_closed/);
    gate.resolve();
    await closing;
    await new Promise((resolve) => setImmediate(resolve));
    assert.match(late.message, /transaction_closed|client_closed/);
    assert.equal(bridge.closed, true);
  });
});

test("the shared envelope fixtures carry the shapes the bridge sends and switches on", () => {
  const inputTypes = new Set(envelopes.inputs.map((input) => input.type));
  assert.deepEqual([...inputTypes].sort(), [
    "callbackResult",
    "close",
    "effectResult",
    "task",
    "transactionCommand",
  ]);
  for (const input of envelopes.inputs) {
    switch (input.type) {
      case "task":
        assert.equal(typeof input.requestId, "string");
        assert.equal(typeof input.command.kind, "string");
        break;
      case "transactionCommand":
        assert.equal(typeof input.requestId, "string");
        assert.equal(typeof input.transactionId, "string");
        assert.ok(input.scope === undefined || typeof input.scope === "string");
        assert.equal(typeof input.command.kind, "string");
        break;
      case "callbackResult":
        assert.equal(typeof input.effectId, "string");
        assert.equal(typeof input.transactionId, "string");
        assert.equal(typeof input.ok, "boolean");
        assert.ok(
          input.ok ? !("error" in input) : typeof input.error === "string",
        );
        break;
      case "effectResult":
        assert.equal(typeof input.effectId, "string");
        assert.equal(typeof input.outcome.ok, "boolean");
        if (!input.outcome.ok) {
          assert.equal(typeof input.outcome.error.message, "string");
          assert.ok(
            input.outcome.error.status === undefined ||
              typeof input.outcome.error.status === "number",
          );
        }
        break;
      case "close":
        assert.deepEqual(Object.keys(input), ["type"]);
        break;
    }
  }
  const eventTypes = new Set(envelopes.events.map((event) => event.type));
  assert.deepEqual([...eventTypes].sort(), [
    "callCompleted",
    "cancelEffect",
    "changed",
    "effect",
    "laneSignal",
    "observerChanged",
    "report",
    "runtimeClosed",
    "taskCompleted",
  ]);
  const operations = new Set();
  const lanes = new Set();
  for (const event of envelopes.events) {
    switch (event.type) {
      case "taskCompleted":
        assert.equal(typeof event.requestId, "string");
        assert.equal(typeof event.ok, "boolean");
        assert.ok("value" in event);
        assert.ok(
          event.ok ? !("error" in event) : typeof event.error === "string",
        );
        break;
      case "effect":
        assert.equal(typeof event.effectId, "string");
        assert.equal(typeof event.operation.kind, "string");
        operations.add(event.operation.kind);
        if (event.operation.kind === "callback") {
          assert.equal(typeof event.operation.transactionId, "string");
          assert.equal(typeof event.operation.requestId, "string");
        }
        break;
      case "cancelEffect":
        assert.equal(typeof event.effectId, "string");
        break;
      case "callCompleted":
        assert.equal(typeof event.callId, "string");
        assert.equal(typeof event.outcome, "object");
        break;
      case "observerChanged":
        assert.equal(typeof event.observerId, "string");
        assert.ok("snapshot" in event);
        break;
      case "report":
        assert.ok(
          ["records", "error", "protocol"].includes(event.diagnostic.kind),
        );
        break;
      case "changed":
        assert.ok(event.tables.every((table) => typeof table === "string"));
        break;
      case "laneSignal":
        assert.equal(typeof event.signal.lane, "string");
        lanes.add(event.signal.lane);
        break;
      case "runtimeClosed":
        assert.deepEqual(Object.keys(event), ["type"]);
        break;
    }
  }
  assert.deepEqual([...operations].sort(), [
    "callback",
    "http",
    "prerequisite",
    "refreshAuth",
    "socket",
    "timer",
  ]);
  // Every DownlinkSignal the subscription projection understands.
  assert.deepEqual([...lanes].sort(), [
    "acknowledged",
    "bootstrap",
    "changed",
    "ended",
    "opened",
    "paused",
    "requests",
    "resumed",
    "stopped",
  ]);
});

test("the bridge dispatches every fixture event and answers effects it has no handler for", async () => {
  // A carrier that delivers the shared fixture events after a successful
  // open: the bridge must route each one, never stop on one, and answer an
  // effect nobody handles.
  let wake;
  let batches = [];
  const submitted = [];
  const detached = [];
  const carrier = {
    runtimeOpen(request, wakeRuntime) {
      const { requestId } = JSON.parse(request);
      wake = wakeRuntime;
      batches.push([
        {
          type: "taskCompleted",
          requestId,
          ok: true,
          value: {
            clientId: "c",
            schema: { rebuilt: false, pending: null, lastRebuild: null },
          },
        },
      ]);
      setImmediate(() => wake("9"));
      return "9";
    },
    runtimeSubmit(runtimeId, message) {
      assert.equal(runtimeId, "9");
      submitted.push(JSON.parse(message));
    },
    runtimeDrain(runtimeId) {
      assert.equal(runtimeId, "9");
      return JSON.stringify(batches.shift() ?? []);
    },
    runtimeDetach(runtimeId) {
      detached.push(runtimeId);
    },
  };
  const { bridge, opened } = await Bridge.open(carrier, {
    path: "unused",
    schema,
  });
  assert.equal(opened.clientId, "c");
  const seen = [];
  for (const type of [
    "callCompleted",
    "observerChanged",
    "report",
    "changed",
    "laneSignal",
  ])
    bridge.on(type, (event) => seen.push(event.type));
  const handled = [];
  bridge.onEffect("timer", (effectId, operation) =>
    handled.push([effectId, operation.millis]),
  );
  // Split the fixture events over two drains to exercise the recheck.
  const events = envelopes.events.filter(
    (event) => event.type !== "runtimeClosed",
  );
  batches = [events.slice(0, 5), events.slice(5)];
  wake("9");
  assert.deepEqual(seen, [
    "callCompleted",
    "observerChanged",
    "report",
    "report",
    "report",
    "changed",
    ...Array(9).fill("laneSignal"),
  ]);
  assert.deepEqual(handled, [["7", 250]]);
  const answered = submitted.filter((input) => input.type !== "callbackResult");
  assert.deepEqual(answered.map((input) => input.effectId).sort(), [
    "10",
    "11",
    "12",
    "6",
    "8",
    "9",
  ]);
  for (const input of answered)
    assert.deepEqual(input.outcome, {
      ok: false,
      error: { message: "unsupported effect" },
    });
  // A callback effect for a request the bridge does not route is refused.
  assert.deepEqual(
    submitted.filter((input) => input.type === "callbackResult"),
    [
      {
        type: "callbackResult",
        effectId: "5",
        transactionId: "tx7",
        ok: false,
        error: "unknown transaction",
      },
    ],
  );
  const task = bridge.task({ kind: "status" });
  const closing = bridge.close();
  assert.deepEqual(submitted.at(-1), { type: "close" });
  batches = [[{ type: "runtimeClosed" }]];
  wake("9");
  await assert.rejects(task, /client_closed/);
  await closing;
  assert.deepEqual(detached, ["9"]);
  assert.equal(bridge.closed, true);
});

test("a throwing listener does not stop the completion in the same drain batch", async () => {
  const original = globalThis.reportError;
  const reported = [];
  globalThis.reportError = (error) => reported.push(error);
  try {
    await withBridge(async (bridge) => {
      const failure = Error("listener failed");
      const seen = [];
      bridge.on("changed", () => {
        throw failure;
      });
      bridge.on("changed", (event) => seen.push(event.tables));
      assert.equal(await bridge.task(create("l", "A")), null);
      assert.deepEqual(reported, [failure]);
      assert.equal(seen.length, 1);
      assert.ok(seen[0].includes("Entry"));
    });
  } finally {
    if (original === undefined) delete globalThis.reportError;
    else globalThis.reportError = original;
  }
});

/** Run a module script and answer its exit code and output. */
function script(source) {
  return new Promise((resolve) => {
    execFile(
      process.execPath,
      ["--input-type=module", "-e", source],
      { timeout: 20000 },
      (error, stdout, stderr) =>
        resolve({
          code: error ? (error.code ?? error.signal) : 0,
          stdout,
          stderr,
        }),
    );
  });
}
const bridgeModule = fileURLToPath(
  new URL("../../../packages/client-js/bridge.mts", import.meta.url),
);
const addon = fileURLToPath(
  new URL("../../../bindings/node/axton-node.node", import.meta.url),
);
const prelude = `
import { createRequire } from "node:module";
import { mkdtempSync, readFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
const { Bridge } = await import(${JSON.stringify(bridgeModule)});
const native = createRequire(import.meta.url)(${JSON.stringify(addon)});
const schema = JSON.parse(readFileSync(${JSON.stringify(fileURLToPath(schemaUrl))}, "utf8"));
const path = join(mkdtempSync(join(tmpdir(), "axton-bridge-exit-")), "client.sqlite");
`;

test("an outstanding task keeps the process alive until it settles", async () => {
  const { code, stdout, stderr } = await script(`${prelude}
const { bridge } = await Bridge.open(native, { path, schema });
await bridge.task(${JSON.stringify(create("x", "kept"))});
const row = await bridge.task(${JSON.stringify(read("x"))});
console.log("read " + row.text);
`);
  assert.equal(code, 0, stderr);
  assert.equal(stdout.trim(), "read kept");
});

test("an idle open client does not keep the process alive", async () => {
  const started = Date.now();
  const { code, stdout, stderr } = await script(`${prelude}
await Bridge.open(native, { path, schema });
console.log("opened");
`);
  assert.equal(code, 0, stderr);
  assert.equal(stdout.trim(), "opened");
  assert.ok(Date.now() - started < 15000);
});

File: packages/dart/test/runtime_bridge_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:axton/axton.dart';
import 'package:axton/src/bridge.dart';
import 'package:test/test.dart';

/// A fresh temporary file with the Entry schema, opened through the real
/// native runtime as a [Client] or a bare [Bridge].
class Fixture {
  Fixture(this.dir, this.schema);
  final Directory dir;
  final Map<String, dynamic> schema;
  String get path => '${dir.path}/db';
  static String get library => Platform.environment['AXTON_LIBRARY']!;

  static Future<Fixture> create(String prefix) async {
    final dir = await Directory.systemTemp.createTemp(prefix);
    final schema =
        jsonDecode(
              await File('../../fixtures/schemas/entry.json').readAsString(),
            )
            as Map<String, dynamic>;
    return Fixture(dir, schema);
  }

  Future<Client> client() =>
      Client.open(path: path, schema: schema, libraryPath: library);
  Future<Bridge> bridge() =>
      Bridge.open(path: path, schema: schema, libraryPath: library);
  Future<void> dispose() => dir.delete(recursive: true);
}

Map<String, dynamic> create(String text) => {
  'model': 'Entry',
  'op': 'create',
  'identity': {'id': 'e'},
  'values': {'text': text},
};

Map<String, dynamic> update(String text) => {
  'model': 'Entry',
  'op': 'update',
  'identity': {'id': 'e'},
  'values': {'text': text},
};

Future<String?> text(Client client) async =>
    (await client.read('Entry', {'id': 'e'}))?['text'] as String?;

/// Whether [future] settles within [wait]: `'settled'` or `'pending'`.
Future<String> settles(Future<Object?> future, [int wait = 50]) => future
    .then<String>((_) => 'settled', onError: (Object _) => 'settled')
    .timeout(Duration(milliseconds: wait), onTimeout: () => 'pending');

void main() {
  late Fixture fixture;
  setUp(() async => fixture = await Fixture.create('axton-bridge-'));
  tearDown(() => fixture.dispose());

  test('overlapping tasks return to the waiter that submitted them', () async {
    final client = await fixture.client();
    try {
      await client.transaction((tx) => tx.direct(create('hello')));
      final gate = Completer<void>();
      final entered = Completer<void>();
      final order = <String>[];
      final transaction = client.transaction((tx) async {
        await tx.direct(update('inside'));
        expect((await tx.read('Entry', {'id': 'e'}))!['text'], 'inside');
        entered.complete();
        await gate.future;
        await tx.direct(update('committed'));
      });
      unawaited(transaction.then((_) => order.add('transaction')));
      await entered.future;
      final read = client.read('Entry', {'id': 'e'});
      final state = client.syncState();
      unawaited(read.then((_) => order.add('read')));
      unawaited(state.then((_) => order.add('syncState')));
      expect(await settles(read), 'pending');
      expect(await settles(state), 'pending');
      gate.complete();
      await transaction;
      expect((await read)!['text'], 'committed');
      expect((await state)['pending'], 0);
      expect(order, ['transaction', 'read', 'syncState']);
      expect(await text(client), 'committed');
    } finally {
      await client.close();
    }
  });

  test('a task after close fails client_closed; close is idempotent', () async {
    final bridge = await fixture.bridge();
    expect(bridge.opened['clientId'], isA<String>());
    await bridge.close();
    await expectLater(
      bridge.task({'kind': 'status'}),
      throwsA(
        isA<StateError>().having((e) => e.message, 'message', 'client_closed'),
      ),
    );
    await bridge.close();
    expect(Bridge.attached, isNot(contains(bridge.runtimeId)));
  });

  test(
    'a throwing callback rethrows the same object and rolls back; a caught failed command fails at commit',
    () async {
      final client = await fixture.client();
      try {
        await client.transaction((tx) => tx.direct(create('hello')));
        final thrown = _Thrown();
        Object? caught;
        try {
          await client.transaction((tx) async {
            await tx.direct(update('rolled back'));
            throw thrown;
          });
        } catch (error) {
          caught = error;
        }
        expect(identical(caught, thrown), isTrue);
        expect(await text(client), 'hello');

        Object? engine;
        Object? failed;
        try {
          await client.transaction((tx) async {
            try {
              await tx.direct(create('duplicate'));
            } catch (error) {
              engine = error;
            }
            await tx.direct(update('after failure'));
          });
        } catch (error) {
          failed = error;
        }
        expect(engine, isA<StateError>());
        expect(
          failed,
          isA<StateError>().having(
            (e) => e.message,
            'message',
            (engine! as StateError).message,
          ),
        );
        expect(await text(client), 'hello');
      } finally {
        await client.close();
      }
    },
  );

  test('nested savepoints carry the scope Rust issued for them', () async {
    final client = await fixture.client();
    try {
      await client.transaction((tx) async {
        await tx.direct(create('top'));
        await tx.savepoint(() async {
          await tx.direct(update('outer'));
          try {
            await tx.savepoint(() async {
              await tx.direct(update('inner'));
              throw StateError('inner');
            });
          } on StateError catch (_) {}
          expect((await tx.read('Entry', {'id': 'e'}))!['text'], 'outer');
          await tx.savepoint(() => tx.direct(update('released')));
        });
        expect((await tx.read('Entry', {'id': 'e'}))!['text'], 'released');
      });
      expect(await text(client), 'released');
    } finally {
      await client.close();
    }
  });

  test('a failed open throws and leaves nothing attached', () async {
    final before = Bridge.attached.toSet();
    await expectLater(
      Bridge.open(
        path: '${fixture.dir.path}/missing/dir/db',
        schema: fixture.schema,
        libraryPath: Fixture.library,
      ),
      throwsStateError,
    );
    expect(Bridge.attached.toSet(), before);
    final client = await fixture.client();
    try {
      await client.transaction((tx) => tx.direct(create('after')));
      expect(await text(client), 'after');
    } finally {
      await client.close();
    }
    expect(Bridge.attached.toSet(), before);
  });

  test('close during an open callback fails its later commands', () async {
    final client = await fixture.client();
    final gate = Completer<void>();
    final entered = Completer<void>();
    final later = Completer<Object>();
    final transaction = client.transaction((tx) async {
      await tx.direct(create('never'));
      entered.complete();
      await gate.future;
      try {
        await tx.read('Entry', {'id': 'e'});
        later.complete('succeeded');
      } catch (error) {
        later.complete(error);
      }
    });
    await entered.future;
    final closing = client.close();
    await expectLater(
      transaction,
      throwsA(
        isA<StateError>().having((e) => e.message, 'message', 'client_closed'),
      ),
    );
    gate.complete();
    expect(
      await later.future,
      isA<StateError>().having(
        (e) => e.message,
        'message',
        anyOf('transaction_closed', 'client_closed'),
      ),
    );
    await closing;
    final reopened = await fixture.client();
    try {
      expect(await text(reopened), isNull, reason: 'the open unit rolled back');
    } finally {
      await reopened.close();
    }
  });

  test('the bridge envelopes match the shared fixtures', () async {
    final fixtures =
        jsonDecode(
              await File('../../fixtures/bridge/envelopes.json').readAsString(),
            )
            as Map<String, dynamic>;
    final inputs = (fixtures['inputs'] as List).cast<Map<String, dynamic>>();
    final events = (fixtures['events'] as List).cast<Map<String, dynamic>>();
    for (final entry in [...inputs, ...events]) {
      expect(entry['type'], isA<String>(), reason: '$entry');
    }
    // Every input the bridge builds has the spelling Rust decodes.
    final built = [
      Bridge.taskEnvelope('42', {
        'kind': 'read',
        'key': {
          'model': 'Todo',
          'identity': {'id': 't'},
        },
      }),
      Bridge.taskEnvelope('44', {'kind': 'transaction'}),
      Bridge.transactionCommandEnvelope('43', 'tx7', 'sp1', {
        'kind': 'direct',
        'operation': {},
      }),
      Bridge.transactionCommandEnvelope('45', 'tx7', null, {
        'kind': 'savepoint',
      }),
      Bridge.callbackResultEnvelope('5', 'tx7', ok: true),
      Bridge.callbackResultEnvelope('5', 'tx7', ok: false, error: 'boom'),
      Bridge.effectResultEnvelope('101', ok: true, value: {'event': 'opened'}),
      Bridge.effectResultEnvelope(
        '101',
        ok: true,
        value: {'event': 'message', 'body': '{}'},
      ),
      Bridge.effectResultEnvelope(
        '101',
        ok: true,
        value: {'event': 'overflow'},
      ),
      Bridge.effectResultEnvelope('101', ok: true, value: {'event': 'closed'}),
      Bridge.effectResultEnvelope(
        '102',
        ok: false,
        error: 'pull failed',
        status: 401,
      ),
      Bridge.effectResultEnvelope('103', ok: false, error: 'offline'),
      Bridge.effectResultEnvelope(
        '104',
        ok: true,
        value: {'status': 200, 'body': '{}'},
      ),
      Bridge.closeEnvelope,
    ];
    expect(jsonDecode(jsonEncode(built)), inputs);
    // Every event carries what the dispatcher switches on.
    final seen = <String>{};
    for (final event in events) {
      final type = event['type'] as String;
      seen.add(type);
      switch (type) {
        case 'taskCompleted':
          expect(event['requestId'], isA<String>());
          expect(event['ok'], isA<bool>());
          expect(event.containsKey('value'), isTrue);
          if (event['ok'] == false) expect(event['error'], isA<String>());
          // A failure's machine-readable reason, when it has one.
          if (event.containsKey('details')) {
            expect((event['details'] as Map)['code'], isA<String>());
          }
        case 'effect':
          expect(event['effectId'], isA<String>());
          final operation = event['operation'] as Map<String, dynamic>;
          expect(operation['kind'], isA<String>());
          if (operation['kind'] == 'callback') {
            expect(operation['transactionId'], isA<String>());
            expect(operation['requestId'], isA<String>());
          }
        case 'cancelEffect':
          expect(event['effectId'], isA<String>());
        case 'callCompleted':
          expect(event['callId'], isA<String>());
          expect(event.containsKey('outcome'), isTrue);
        case 'observerChanged':
          expect(event['observerId'], isA<String>());
          expect(event.containsKey('snapshot'), isTrue);
        case 'report':
          expect((event['diagnostic'] as Map)['kind'], isA<String>());
        case 'changed':
          expect((event['tables'] as List).cast<String>(), isNotEmpty);
        case 'runtimeClosed':
          break;
        default:
          fail('unknown event type $type');
      }
    }
    expect(seen, {
      'taskCompleted',
      'effect',
      'cancelEffect',
      'callCompleted',
      'observerChanged',
      'report',
      'changed',
      'runtimeClosed',
    });
  });

  test(
    'a throwing changed listener cannot stop the completion in its batch',
    () async {
      final bridge = await fixture.bridge();
      try {
        final reported = <Object>[];
        final done = Completer<Object?>();
        runZonedGuarded(() {
          bridge.changed.listen((_) => throw StateError('listener'));
          bridge
              .task({'kind': 'direct', 'operation': create('hello')})
              .then(done.complete, onError: done.completeError);
        }, (error, _) => reported.add(error));
        await done.future.timeout(const Duration(seconds: 5));
        expect(reported, [
          isA<StateError>().having((e) => e.message, 'message', 'listener'),
        ]);
        final row = await bridge.task({
          'kind': 'read',
          'key': {
            'model': 'Entry',
            'identity': {'id': 'e'},
          },
        });
        expect((row as Map)['text'], 'hello');
      } finally {
        await bridge.close();
      }
    },
  );

  test(
    'an observer claimed at its task completion hears the snapshot of the same batch',
    () async {
      final bridge = await fixture.bridge();
      try {
        final heard = <Map<String, dynamic>>[];
        String? claimed;
        final value = await bridge.task(
          {'kind': 'scopeSubscribe', 'scope': 'book'},
          onValue: (value) {
            claimed = (value as Map)['observerId'] as String;
            expect(heard, isEmpty, reason: 'claimed before its first snapshot');
            bridge.listen(claimed!, heard.add);
          },
        );
        expect((value as Map)['observerId'], claimed);
        expect(heard, hasLength(1), reason: 'the first snapshot was not lost');
        expect(heard.single['kind'], 'subscription');
        expect((heard.single['status'] as Map)['connection'], 'offline');
        // The runtime's close ends the observer with its terminal snapshot.
        await bridge.close();
        expect(heard, hasLength(2));
        expect(heard.last['closed'], isTrue);
      } finally {
        await bridge.close();
      }
    },
  );

  test('a failure with a code carries it beside the message', () async {
    final bridge = await fixture.bridge();
    try {
      await expectLater(
        bridge.task({
          'kind': 'scopeBootstrap',
          'scope': 'book',
          'subscriptionId': 99,
        }),
        throwsA(
          isA<TaskFailure>()
              .having(
                (e) => e.message,
                'message',
                startsWith('subscription.closed'),
              )
              .having((e) => e.details, 'details', {
                'code': 'subscription.closed',
              }),
        ),
      );
      await expectLater(
        bridge.task({'kind': 'nope'}),
        throwsA(
          isA<StateError>().having(
            (e) => e is TaskFailure ? e.details : null,
            'details',
            isNull,
          ),
        ),
      );
    } finally {
      await bridge.close();
    }
  });

  test('a malformed envelope is reported and completes nothing', () async {
    final bridge = await fixture.bridge();
    try {
      final report = bridge.reports.first;
      bridge.submitRaw({'type': 'nope'});
      final diagnostic = await report.timeout(const Duration(seconds: 5));
      expect(diagnostic['kind'], 'protocol');
      expect(await bridge.task({'kind': 'status'}), isA<Map>());
    } finally {
      await bridge.close();
    }
  });
}

class _Thrown {}
